gene_expression_processing.py

```diff
@@ -0,0 +1,272 @@
+    expression_df = pd.DataFrame(
+        data=gene_mat.T, index=gene_adata.raw.var.index, columns=gene_adata.obs.index
+    )
+    pheno_cov_df = pd.concat([expression_df, cov_df, smf_df], axis=1)
+
+    with to_path(out_path).open('w') as pcf:
+        pheno_cov_df.to_csv(pcf, index=False)
+
+
+def get_gene_cis_file(gene_info_df, gene: str, window_size: int):
+    """Get gene cis window file"""
+    # select the gene from df
+    gene_info_gene = gene_info_df[gene_info_df['gene'] == gene]
+    # get gene chromosome
+    chrom = gene_info_gene['chr']
+    # get gene body position (start and end) and add window
+    left_boundary = max(1, int(gene_info_gene['start']) - window_size)
+    right_boundary = min(
+        int(gene_info_gene['end']) + window_size,
+        hl.get_reference('GRCh38').lengths[chrom],
+    )
+    data = {'chromosome': chrom, 'start': left_boundary, 'end': right_boundary}
+    # check if I need an index at all
+    return pd.DataFrame(data, index=[gene])
+
+
+@click.command()
+@click.option('--celltypes')
+@click.option('--chromosomes')
+@click.option('--gene-info-tsv')
+@click.option('--expression-files-prefix')
+@click.option('--sample-mapping-file-path')
+@click.option('--min-pct-expr')
+@click.option('--cis-window-size')
+@click.option(
+    '--max-gene-concurrency',
+    type=int,
+    default=50,
+    help=(
+        'To avoid resource starvation, set this concurrency to limit '
+        'horizontal scale. Higher numbers have a better walltime, but '
+        'risk jobs that are stuck (which are expensive)'
+    ),
+)
+def expression_pipeline(
+    celltypes: str,
+    chromosomes: str,
+    gene_info_tsv: str,
+    expression_files_prefix: str,
+    sample_mapping_file_path: str,
+    min_pct_expr: int = 5,
+    cis_window_size: int = 100000,
+    max_gene_concurrency=100,
+):
+    """
+    Run expression processing pipeline
+    """
+
+    logging.info(f'Cell types to run: {celltypes}')
+    logging.info(f'Chromosomes to run: {chromosomes}')
+
+    # create phenotype covariate files
+    smf_df = pd.read_csv(sample_mapping_file_path, sep='\t')
+    gene_info_df = pd.read_csv(gene_info_tsv, sep='\t')
+
+    # Setup MAX concurrency by genes
+    _dependent_jobs: list[hb.batch.job.Job] = []
+
+    def manage_concurrency_for_job(job: hb.batch.job.Job):
+        """
+        To avoid having too many jobs running at once, we have to limit concurrency.
+        """
+        if len(_dependent_jobs) >= max_gene_concurrency:
+            job.depends_on(_dependent_jobs[-max_gene_concurrency])
+        _dependent_jobs.append(job)
+
+    for celltype in celltypes.split(','):
+        # get covariates (cell type specific)
+        cov_df = get_celltype_covariates(
+            expression_files_prefix=expression_files_prefix, cell_type=celltype
+        )
+        for chromosome in chromosomes.split(','):
+            # get expression (cell type + chromosome)
+            expr_adata: sc.AnnData = get_chrom_celltype_expression(
```

Comment:
You're completing a lot of loading + compute inside loops.

Can you do all this loading, filtering and deciding what to run outside a loop? Or, is there a way to look at doing this in parallel, because you're paying a sequential and high IO tax lots of times.